File: testable-agent/src/test/java/com/alibaba/testable/agent/util/PathUtilTest.java
package com.alibaba.testable.agent.util;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class PathUtilTest {

    @Test
    void should_get_sub_folder() {
        assertEquals("/ab/cd", PathUtil.getFirstLevelFolder("/ab", "/ab/cd/ef/gh"));
        assertEquals("c:\\ab\\cd", PathUtil.getFirstLevelFolder("c:\\ab", "c:\\ab\\cd\\ef\\gh"));
        assertEquals("", PathUtil.getFirstLevelFolder("/ab", "/ab"));
        assertEquals("", PathUtil.getFirstLevelFolder("/ab/cd", "/ab"));
    }

    @Test
    void should_distinguish_absolute_and_relative_path() {
        assertTrue(PathUtil.isAbsolutePath("/a/b/c"));
        assertTrue(PathUtil.isAbsolutePath("c:\\a\\b\\c"));
        assertFalse(PathUtil.isAbsolutePath("a/b/c"));
        assertFalse(PathUtil.isAbsolutePath("a\\b\\c"));
    }

}

File: demo/kotlin-demo/src/test/kotlin/com/alibaba/testable/demo/util/PathUtilTest.kt
package com.alibaba.testable.demo.util

import org.junit.jupiter.api.Test
import com.alibaba.testable.core.annotation.TestableMock
import com.alibaba.testable.core.tool.TestableTool.verify
import java.io.File

class PathUtilTest {

    @TestableMock
    fun exists(f: File): Boolean {
        return when (f.absolutePath) {
            "/a/b" -> true
            "/a/b/c" -> true
            else -> f.exists()
        }
    }

    @TestableMock
    fun isDirectory(f: File): Boolean {
        return when (f.absolutePath) {
            "/a/b/c" -> true
            else -> f.isDirectory
        }
    }

    @TestableMock
    fun delete(f: File): Boolean {
        return true
    }

    @TestableMock
    fun listFiles(f: File): Array<File>? {
        return when (f.absolutePath) {
            "/a/b" -> arrayOf(File("/a/b/c"), File("/a/b/d"))
            "/a/b/c" -> arrayOf(File("/a/b/c/e"))
            else -> f.listFiles()
        }
    }

    @Test
    fun should_able_to_mock_java_method_invoke_in_kotlin() {
        PathUtil.deleteRecursively(File("/a/b/"))
        verify("listFiles").times(2)
        verify("delete").times(4)
    }

}
